Microsoft.Azure.Cosmos/src/Routing/AsyncCacheNonBlocking.cs

```
@@ -386,14 +419,39 @@ public bool ShouldRemoveFromCacheThreadSafe()
                 }
             }
 
+            public Task RefreshAsync(
+                Func<T, Task<T>> createRefreshTask)
+            {
+                lock (this.refreshLock)
+                {
+                    if (AsyncLazyWithRefreshTask<T>.IsTaskRunning(this.backgroundRefreshTask))
+                    {
+                        DefaultTrace.TraceInformation(
+                            message: "Background refresh task is already in progress, skip creating a new one.");
+
+                        return Task.CompletedTask;
+                    }
+                    else
+                    {
+                        DefaultTrace.TraceInformation(
+                            message: "Started a new background refresh task.");
+
+                        this.backgroundRefreshTask = Task.Run(async () => await this.CreateAndWaitForBackgroundRefreshTaskAsync(
```

Any exception in CreateAndWaitForBackgroundRefreshTaskAsync would result in process crash if validation for unobserved ewxceptions is enabled (like in Compute). Have you double-checked that CreateAndWaitForBackgroundRefreshTaskAsync is guaranteed to never throw an exception?

---

I do see that there are possibility this method `CreateAndWaitForBackgroundRefreshTaskAsync ` might throw exceptions. However, this is applicable for `GetAsync()` call as well and it's been there since long time. Do you suggest we should swallow any exceptions to avoid process crash ?

---

So @FabianMeiswinkel  - Right now, I do have a `ContinueWith()` token that avoids to throw any unobserved execptions. I think this should cover us with the process crash.